openfe/protocols/openmm_septop/base.py

```diff
@@ -0,0 +1,1436 @@
+        reporter : multistate.MultiStateReporter
+          The reporter for the simulation.
+        """
+        mdt_top = mdt.Topology.from_openmm(topology)
+
+        selection_indices = mdt_top.select(
+                output_settings.output_indices
+        )
+
+        nc = self.shared_basepath / output_settings.output_filename
+        chk = output_settings.checkpoint_storage_filename
+        chk_intervals = settings_validation.convert_checkpoint_interval_to_iterations(
+            checkpoint_interval=output_settings.checkpoint_interval,
+            time_per_iteration=simulation_settings.time_per_iteration,
+        )
+
+        if output_settings.positions_write_frequency is not None:
+            pos_interval = settings_validation.divmod_time_and_check(
+                numerator=output_settings.positions_write_frequency,
+                denominator=simulation_settings.time_per_iteration,
+                numerator_name="output settings' position_write_frequency",
+                denominator_name="simulation settings' time_per_iteration"
+            )
+        else:
+            pos_interval = 0
+
+        if output_settings.velocities_write_frequency is not None:
+            vel_interval = settings_validation.divmod_time_and_check(
+                numerator=output_settings.velocities_write_frequency,
+                denominator=simulation_settings.time_per_iteration,
+                numerator_name="output settings' velocity_write_frequency",
+                denominator_name="simulation settings' time_per_iteration"
+            )
+        else:
+            vel_interval = 0
+
+        reporter = multistate.MultiStateReporter(
+            storage=nc,
+            analysis_particle_indices=selection_indices,
+            checkpoint_interval=chk_intervals,
+            checkpoint_storage=chk,
+            position_interval=pos_interval,
+            velocity_interval=vel_interval,
+        )
+
+        # Write out the structure's PDB whilst we're here
+        if len(selection_indices) > 0:
+            traj = mdt.Trajectory(
+                positions[selection_indices, :],
+                mdt_top.subset(selection_indices),
+            )
+            traj.save_pdb(
+               self.shared_basepath / output_settings.output_structure
+            )
+
+        return reporter
+
+    def _get_ctx_caches(
+        self,
+        engine_settings: OpenMMEngineSettings
+    ) -> tuple[openmmtools.cache.ContextCache, openmmtools.cache.ContextCache]:
+        """
+        Set the context caches based on the chosen platform
+
+        Parameters
+        ----------
+        engine_settings : OpenMMEngineSettings,
+
+        Returns
+        -------
+        energy_context_cache : openmmtools.cache.ContextCache
+          The energy state context cache.
+        sampler_context_cache : openmmtools.cache.ContextCache
+          The sampler state context cache.
+        """
+        platform = omm_compute.get_openmm_platform(
```

Comment:
I would copy over the restrict_cpu code from here: https://github.com/OpenFreeEnergy/openfe/blob/59811f647178f583133fb62ef97da8ae682026e4/openfe/protocols/openmm_afe/base.py#L792